File: python/scripts/verify_reasoning_order.py
# FILE: python/scripts/verify_reasoning_order.py
"""
Introspects the FastMCP server's advertised tool schemas and asserts that
EVERY tool declares `reasoning` as (a) the first property and (b) a required
field of type string.

Run: uv run python scripts/verify_reasoning_order.py
Exit 0 = all good, 1 = at least one violation.
"""

import asyncio
import sys


def _extract_schema(tool) -> dict:
    """
    FastMCP Tool objects expose their JSON schema under different attribute
    names across versions (parameters / inputSchema / input_schema / schema).
    Probe them in order and return the first dict that has 'properties'.
    """
    for attr in ("parameters", "inputSchema", "input_schema", "schema"):
        val = getattr(tool, attr, None)
        if isinstance(val, dict) and "properties" in val:
            return val
    # Fall back: some versions nest it, or expose a model with .model_json_schema()
    for attr in ("parameters", "inputSchema", "input_schema"):
        val = getattr(tool, attr, None)
        if val is not None and hasattr(val, "model_json_schema"):
            return val.model_json_schema()
    return {}


async def main() -> int:
    from adeu.server import mcp

    tools = await mcp.list_tools()
    if not tools:
        print("❌ list_tools returned no tools.", file=sys.stderr)
        return 1

    # Debug aid: on first run, dump the attribute names + schema of one tool so
    # we can confirm the introspection is reading the right field.
    if "--debug" in sys.argv:
        t0 = tools[0]
        print(f"[debug] tool[0]={t0.name}", file=sys.stderr)
        print(
            f"[debug] dir()={[a for a in dir(t0) if not a.startswith('_')]}",
            file=sys.stderr,
        )
        print(f"[debug] extracted schema={_extract_schema(t0)}", file=sys.stderr)

    failures = 0
    for tool in tools:
        schema = _extract_schema(tool)
        props = schema.get("properties", {})
        required = schema.get("required", [])
        keys = list(props.keys())

        first_key = keys[0] if keys else None
        is_first = first_key == "reasoning"
        is_required = "reasoning" in required
        is_string = isinstance(props.get("reasoning"), dict) and props["reasoning"].get("type") == "string"

        if is_first and is_required and is_string:
            print(f"✅ {tool.name}: reasoning is first + required (string)")
        else:
            failures += 1
            print(
                f"❌ {tool.name}: reasoning check failed "
                f"(first_key={first_key}, required={is_required}, string={is_string})",
                file=sys.stderr,
            )

    print(f"\n{len(tools) - failures}/{len(tools)} tools passed.", file=sys.stderr)
    return 0 if failures == 0 else 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: node/packages/mcp-server/scripts/verify-reasoning-order.mjs
// FILE: node/packages/mcp-server/scripts/verify-reasoning-order.mjs
// Boots the compiled MCP server over stdio, runs initialize + tools/list, and
// asserts that EVERY tool declares `reasoning` as (a) the first property and
// (b) a required field. Exit code 0 = all good, 1 = at least one violation.
//
// Usage: node scripts/verify-reasoning-order.mjs
// Requires: npm run build (dist/index.js must exist).

import { spawn } from "node:child_process";
import { resolve, dirname } from "node:path";
import { existsSync } from "node:fs";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const serverPath = resolve(__dirname, "../dist/index.js");

if (!existsSync(serverPath)) {
  console.error(
    `❌ Server not built: ${serverPath}. Run 'npm run build' first.`,
  );
  process.exit(1);
}

const proc = spawn("node", [serverPath]);
const pending = new Map();
let rpcId = 100;
let buf = "";

proc.stdout.on("data", (data) => {
  buf += data.toString();
  let idx;
  while ((idx = buf.indexOf("\n")) !== -1) {
    const line = buf.slice(0, idx).trim();
    buf = buf.slice(idx + 1);
    if (!line.startsWith("{")) continue;
    try {
      const msg = JSON.parse(line);
      if (msg.id !== undefined && pending.has(msg.id)) {
        const cb = pending.get(msg.id);
        pending.delete(msg.id);
        cb(msg);
      }
    } catch {
      /* ignore partial/non-JSON */
    }
  }
});

function rpc(method, params) {
  const id = ++rpcId;
  return new Promise((res, rej) => {
    const t = setTimeout(() => rej(new Error(`RPC timeout: ${method}`)), 15000);
    pending.set(id, (m) => {
      clearTimeout(t);
      res(m);
    });
    proc.stdin.write(
      JSON.stringify({ jsonrpc: "2.0", id, method, params }) + "\n",
    );
  });
}

function notify(method, params) {
  proc.stdin.write(JSON.stringify({ jsonrpc: "2.0", method, params }) + "\n");
}

try {
  await rpc("initialize", {
    protocolVersion: "2024-11-05",
    capabilities: {},
    clientInfo: { name: "verify-reasoning", version: "0.0.0" },
  });
  notify("notifications/initialized", {});

  const list = await rpc("tools/list", {});
  const tools = list.result?.tools ?? [];

  if (tools.length === 0) {
    console.error("❌ tools/list returned no tools.");
    process.exit(1);
  }

  let failures = 0;
  for (const tool of tools) {
    const schema = tool.inputSchema ?? {};
    const props = schema.properties ?? {};
    const keys = Object.keys(props);
    const required = schema.required ?? [];

    const firstKey = keys[0];
    const isFirst = firstKey === "reasoning";
    const isRequired = required.includes("reasoning");
    const isString = props.reasoning?.type === "string";

    if (isFirst && isRequired && isString) {
      console.log(`✅ ${tool.name}: reasoning is first + required (string)`);
    } else {
      failures++;
      console.error(
        `❌ ${tool.name}: reasoning check failed ` +
          `(firstKey=${firstKey}, required=${isRequired}, string=${isString})`,
      );
    }
  }

  console.error(`\n${tools.length - failures}/${tools.length} tools passed.`);
  proc.kill();
  process.exit(failures === 0 ? 0 : 1);
} catch (e) {
  console.error(`❌ ${e.message}`);
  proc.kill();
  process.exit(1);
}
